When verifying a nonce, an empty referrer should be allowed. even though expectedReferrerHost is specified

kl-he opened the pull request:
### Description:

if an expectedReferrerHost is passed, the nonce-check will fail, if there is no referrer. This was introduced in #17228, as a fix when using GoogleAnalyticsImporter.
While using GoogleAnalyticsImporter, we are seeing a referrer being sent by the browser, so this this will cause the check to fail, as the empty referrer doesn't match google.com.

This fix allows empty referrers again, while also doing expectedReferrerHost and/or localhost, when referrer is specified.

Tests added as integration tests.
./console tests:run --testsuite integration --filter NonceTest

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

klh-br committed:
Allow empty referrer with expected referrer set

Expected referrer would fail a nonce-check, if the referrer was empty or
local. Introduced in 715ded1a/#17228.
This fix "reverts" to a pure nonce-check, if referrer is empty or
local.

Issue encountered using the the GoogleAnalyticsImporter, also referred
to in #17228.

klh-br committed:
Fixed misunderstanding with localhost-handling

Previous fix would allow localhost, even though expectedReferrerHost
specified something else.

If a referrer is present, it must fail if it is not localhost or
expectedReferrerHost.

klh-br committed:
Updated some comments to match what is happening

sgiehl commented:
Hi @kl-he 
Thanks for contributing this PR. That's very appreciated. 
The code imho looks fine and it's awesome you also added some tests.

I'm not 100% sure about possible security implications of removing that check. In theory that would now allow to simply bypass the expected referrer check by suppressing the referrer header.
On the other side I guess there might be browsers or addons, that suppress that header on purpose for privacy reasons maybe...
Looking at the code before https://github.com/matomo-org/matomo/pull/17228, it looks like empty referrer were accepted as well. So might be fine to simply merge this one.

@tsteur Any thoughts or objections on that one?

tsteur commented:
@kl-he could you maybe explain the use case a bit more why you need the change? This will help understand things.

kl-he commented:
We are in the opposite situation as [diosmosis](https://github.com/diosmosis) was in, when #17228 was submitted.

We are not seeing a referrer being sent (tried with different browsers and distros), when authenticating the Google Analytics Importer plugin, resulting in a failed authorization attempt.

Without #17228 we would not have had a problem, as it allowed empty referrers, as @sgiehl correctly points out.

This pull-request will allow the empty referrers again, but still do the domain-check, if a referrer is passed.

If you go ahead with this pull-request, it might make sense to rename $expectedReferrerHost to $allowedReferrerHost.

sgiehl commented:
@tsteur any further notes on this one? I'm otherwise tempted to merge it.

tsteur commented:
@sgiehl figured it's best to do a search on OWASP's thoughts on this so checked https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#verifying-origin-with-standard-headers and https://owasp.org/www-community/attacks/csrf

Maybe there's a redirect that gets rid of the referrer. 
So it seems fine to merge. However, as suggested by @kl-he we could indeed rename the variable to like `allowedReferrerHost` and we should make clear in the comments for the method how it behaves and that this is expected behaviour that when no referrer is there that it's not checked.

sgiehl commented:
@kl-he could you maybe apply those suggestions

klh-br committed:
Renamed expectedReferrer to allowedReferrer

Renamed expectedReferrerHost to allowedReferrerHost.
Also updated doc-blocks and comments, to reflect the changes.

klh-br committed:
Updated comments

klh-br committed:
Merged 4.x-dev

kl-he commented:
@sgiehl Done

sgiehl commented on tests/PHPUnit/Integration/NonceTest.php:
```diff
@@ -0,0 +1,96 @@
+
+    public function testVerifyNonceWithErrorMessage_validNonceAndLocalReferrerWithNoAllowedReferrer_expectEmptyString()
+    {
+        $this->setReferrer('http://app'); // The "local" host when running via CLI.
```
I guess this test is meant to test with the configured hostname as referrer. Might be better to fetch the host with `Config::getHostname()` so it works on all instances.
```suggestion
        $this->setReferrer('http://' . \Piwik\Config::getHostname()); // The "local" host when running via CLI.
```

sgiehl reviewed (changes requested):
Left a comment regarding a failing test. Otherwise this looks good to me
@kl-he could you adjust the test and resolve the conflicts with or branches, then I'll try to merge this one before the next release.

MatomoForumNotifications commented:
This pull request has been mentioned on **Matomo forums**. There might be relevant details there:

https://forum.matomo.org/t/google-analytics-importer/48279/2

sgiehl commented:
Not mergable anymore due to conflicts. Created #20479 to replace this PR.
